lib/Auth/Process/PerunEntitlementExtended.php

```diff
@@ -0,0 +1,148 @@
+        );
+
+        $interface = $configuration->getValueValidate(
+            self::INTERFACE_PROPNAME,
+            [Adapter::RPC, Adapter::LDAP],
+            Adapter::RPC
+        );
+        $this->adapter = Adapter::getInstance($interface);
+    }
+
+    public function process(&$request)
+    {
+        $eduPersonEntitlementExtended = [];
+        $capabilities = [];
+        $forwardedEduPersonEntitlement = [];
+
+        if (isset($request['perun']['groups'])) {
+            $eduPersonEntitlementExtended = $this->getEduPersonEntitlementExtended($request);
+
+            $capabilities = EntitlementUtils::getCapabilities(
+                $request,
+                $this->adapter,
+                $this->entitlementPrefix,
+                $this->entitlementAuthority
+            );
+        } else {
+            Logger::debug(
+                'perun:PerunEntitlementExtended: There are no user groups assigned to facility.' .
+                '=> Skipping getEduPersonEntitlementExtended and getCapabilities'
+            );
+        }
+
+        if ($this->releaseForwardedEntitlement) {
+            $forwardedEduPersonEntitlement = EntitlementUtils::getForwardedEduPersonEntitlement(
+                $request,
+                $this->adapter,
+                $this->forwardedEduPersonEntitlement
+            );
+        }
+
+        $request['Attributes'][$this->outputAttrName] = array_unique(array_merge(
+            $eduPersonEntitlementExtended,
+            $forwardedEduPersonEntitlement,
+            $capabilities
+        ));
+    }
+
+    private function getEduPersonEntitlementExtended(&$request)
+    {
+        $eduPersonEntitlementExtended = [];
+
+        $groups = $request['perun']['groups'];
+        foreach ($groups as $group) {
+            array_push($eduPersonEntitlementExtended, $this->groupNameWrapper($group->getUuid()));
+
+            $groupName = $group->getUniqueName();
+            $groupName = preg_replace('/^(\w*)\:members$/', '$1', $groupName);
+
+            $entitlement = $this->groupEntitlementWrapper($group->getUuid());
+            array_push($eduPersonEntitlementExtended, $entitlement);
+
+            $entitlementWithAttributes = $this->groupEntitlementWithAttributesWrapper($group->getUuid(), $groupName);
+            array_push($eduPersonEntitlementExtended, $entitlementWithAttributes);
+        }
+
+        natsort($eduPersonEntitlementExtended);
+        return $eduPersonEntitlementExtended;
+    }
+
+    private function groupNameWrapper($groupName)
+    {
+        return $this->entitlementPrefix . 'group:' . EntitlementUtils::encodeName($groupName) .
+            '#' . $this->entitlementAuthority;
+    }
+
+    private function groupEntitlementWrapper($groupName)
+    {
+        return $this->entitlementPrefix . 'group:' . $groupName . '#' . $this->entitlementAuthority;
+    }
+
+    private function groupEntitlementWithAttributesWrapper($uuid, $groupName)
```

Comment:
I think these wrapper methods should be extracted into the "utils" class as well as in the case of the capabilities wrapper.